torch/utils/_sympy/functions.py

```diff
@@ -559,26 +570,43 @@ def eval(cls, base, shift):
 
 
 class MinMaxBase(Expr, LatticeOp):  # type: ignore[misc]
-    def __new__(cls, *args, **assumptions):
+    def __new__(cls, *original_args, **assumptions):
         from sympy.core.parameters import global_parameters
 
-        evaluate = assumptions.pop("evaluate", global_parameters.evaluate)
-        args = (sympify(arg) for arg in args)
+        import torch._dynamo.config as config
 
-        # first standard filter, for cls.zero and cls.identity
-        # also reshape Max(a, Max(b, c)) to Max(a, b, c)
+        evaluate = assumptions.pop("evaluate", global_parameters.evaluate)
+        args = (sympify(arg) for arg in original_args)
 
         if evaluate:
             try:
+                # first standard filter, for cls.zero and cls.identity
+                # also reshape Max(a, Max(b, c)) to Max(a, b, c)
                 args = frozenset(cls._new_args_filter(args))  # type: ignore[assignment]
             except ShortCircuit:
                 return cls.zero  # type: ignore[attr-defined]
-            # remove redundant args that are easily identified
-            args = cls._collapse_arguments(args, **assumptions)
-            # find local zeros
-            args = cls._find_localzeros(args, **assumptions)
 
-        args = frozenset(args)
+            # When this is not none we can bypass _find_localzeros and _collapse_arguments.
+            unique_summation_atoms = cls._should_optimize_max_of_unique_summations(
+                original_args
+            )
+
+            if unique_summation_atoms is None:
+                # remove redundant args that are easily identified
+                args = cls._collapse_arguments(args, **assumptions)
+
+                # find local zeros
+                args = cls._find_localzeros(args, **assumptions)
+                args = frozenset(args)  # type: ignore[assignment]
+
+            elif config.run_extra_validations:
+                # Test that applying _collapse_arguments and _find_localzeros do not change the result.
+                args_test = cls._collapse_arguments(args, **assumptions)
+                args_test = cls._find_localzeros(args, **assumptions)
+                args_test = frozenset(args)
+                assert Expr.__new__(
+                    cls, *ordered(args_test), **assumptions
+                ) == Expr.__new__(cls, *ordered(args), **assumptions)
```

Comment:
nit: print the before after expr when this assert fails

Reply:
More broadly, to be honest, I'm not really sure how useful this assert actually will be. At the slightest whiff of trouble, where this assert fails but the results are still semantically correct, I will probably just dump this assert.

Reply:
i will just remove it